src/Engine/Plugins/EngineTracer.cpp

```diff
@@ -0,0 +1,142 @@
+#include "EngineTracer.h"
+
+#include <cassert>
+#include <utility>
+#include <filesystem>
+
+#include "Utility/ScopeGuard.h"
+#include "Utility/Format.h"
+
+#include "Library/Trace/PaintEvent.h"
+#include "Library/Application/PlatformApplication.h"
+#include "Library/Random/Random.h"
+
+#include "Engine/Engine.h"
+
+#include "Application/GameKeyboardController.h" // TODO(captainurist): Engine -> Application dependency
+
+#include "EngineController.h"
+#include "EngineControlPlugin.h"
+#include "EngineTracePlugin.h"
+#include "EngineDeterministicPlugin.h"
+
+EngineTracer::EngineTracer(Options options): _options(options) {}
+
+EngineTracer::~EngineTracer() {
+    assert(!application()); // We're uninstalled.
+}
+
+void EngineTracer::startTraceRecording(const std::string &savePath, const std::string &tracePath) {
+    assert(_options & ENABLE_RECORDING);
+    assert(_state == CHILLING);
+
+    _state = RECORDING;
+    _saveFilePath = savePath;
+    _traceFilePath = tracePath;
+
+    // TODO(captainurist): this is actually unsafe, we need to cancel running routines in dtor if there are any.
+    _controlPlugin->runControlRoutine([this, savePath] (EngineController *game) {
+        _oldFpsLimit = engine->config->graphics.FPSLimit.Get();
+        engine->config->graphics.FPSLimit.Set(1000); // Load game real quick!
+        game->saveGame(savePath);
+        _deterministicPlugin->enterDeterministicMode();
+        game->loadGame(savePath);
+        _deterministicPlugin->resetDeterministicState();
+        _keyboardController->reset(); // Reset all pressed buttons.
+        _tracePlugin->start();
+        engine->config->graphics.FPSLimit.Set(EngineDeterministicPlugin::TARGET_FPS); // But don't turn the party into a wall-running doomguy.
+    });
+}
+
+void EngineTracer::finishTraceRecording() {
+    assert(_options & ENABLE_RECORDING);
+    assert(_state == RECORDING);
+
+    // The code below needs to go into a control routine because it should be run AFTER the previous control routine
+    // is finished, and it could take a while for it to finish.
+    _controlPlugin->runControlRoutine([this, tracePath = _traceFilePath, savePath = _saveFilePath] (EngineController *game) {
+        assert(_tracePlugin->isTracing()); // Just a sanity check.
+        _deterministicPlugin->leaveDeterministicMode();
+        engine->config->graphics.FPSLimit.Set(_oldFpsLimit);
+
+        EventTrace trace = _tracePlugin->finish();
+        trace.header.saveFileSize = std::filesystem::file_size(savePath); // This function can throw.
```

Comment:
Can you throw the save path into the logger here. "Trace saved to: ..."

Reply:
Done.